### PR TITLE
Changelings can no longer benefit from absorbing monkeymen, nor can vampires gain blood from sucking monkeymen off.

### LINKED ISSUE #7637: Potential solution for changeling/vampire monkeyman exploitation
This was mentioned in the thread earlier today. While I don't know how it would be coded, it seems like it would be a more effective fix than simply making catatonics or people not on the crew manifest yield no genomes or usable blood.

*Absorb DNA/suck blood has to have a proc to see if the body hosted a ckey
*If null, no genomes or blood is gained
*If true, genomes/blood is gained normally.

I think this should address all the complaints about lings and vamps doing no actual antagonizing by forcing them to attack other players instead of monkeymen.

### PR DESCRIPTION
**Pull request by icantthinkofanameritenow**
Let's be blunt, there's no reason why something like this should have been out of the game for this long. If it works, it will singlehandedly fix all of the worst problems that plague the changeling and vampire modes. I'll be testing it now to see how well it works- or at least if it prevents absorption of mindless human mobs

Edit: Fully tested, and works exactly as intended. And closes #7637

### CODE REVIEW

**Comment by ghost** [comment]
Long overdue, cowardly antagonists who try to occupy Genetics to get their free level-ups got tiring after a while

Now, we just need to make Changeling and Vampire not awful, which is likely a much harder and long-term project

File: code/game/gamemodes/changeling/changeling_powers.dm [hunk 1]
```diff
@@ -210,6 +210,10 @@
 		to_chat(src, "<span class='warning'>This creature's DNA is ruined beyond useability!</span>")
 		return
 
+	if (mob.mind == null)
```
Comment (ghost) [inline code/game/gamemodes/changeling/changeling_powers.dm]: Should be if(!T.mind) (see above typecast)

File: code/game/gamemodes/vampire/vampire.dm [hunk 2]
```diff
@@ -388,6 +388,9 @@ You are weak to holy things and starlight. Don't go into space and avoid the Cha
 			to_chat(src, "<span class='warning'>Not a drop of blood here</span>")
 			src.mind.vampire.draining = null
 			return 0
+		if (mob.mind == null)
```
Comment (ghost) [inline code/game/gamemodes/vampire/vampire.dm]: Should be if(!H.mind) (see above proc declaration)

**Comment by Konater** [comment]
It's a start

**Comment by 9600bauds** [comment]
Needs to be done in a way that ghosting out doesn't deny the antag from their blood/genomes, does this work that way?
Reminder that if you need to test things that require two players, you can uncomment your ALLOW_GUESTS config, then join your server, exit BYOND, and join your server again

**Comment by icantthinkofanameritenow** [comment]
@9600bauds 
PJB claimed that this would be sufficient, though I'm not 100% sure myself. I guess I could try the testing idea you mentioned, though.

**Comment by icantthinkofanameritenow** [comment]
Tested the ling functions, and it works exactly as intended. Doing it again for vamp now.

**Comment by Imagebard** [comment]
>if(!T.mind)

I have a few questions about this. Can I just ghost to avoid giving a vampire blood? How does Mind work anyways, is it that thing that tracks someone's job?

**Comment by icantthinkofanameritenow** [comment]
@Imagebard

Unlike ckeys, if a mob has a mind it keeps it even after ghosting- I believe it tracks if the mob was ever controlled by a player at any point in time. I tested just now- ghosting from a mob I controlled and respawning as a ling let me absorb the catatonic body normally, but doing it with a monkeyman had no effect at all.

**Comment by ghost** [comment]
@Imagebard When you ghost, your mind is still attached to your body (which is why you can return to your body when being revived without a kerfuffle, for example). It even handles your ckey (when you ghost, you get an abstract ckey like "@dylanstrategie" to signal who was in the body) which persists even if you DC and RC

Only way to obtain a mob without a mind is to never have had a player in it at any point. This only concerns admin spawns that were never controlled and human monkeys

**Comment by kol1th** [comment]
yes thank you

File: html/changelogs/icantthinkofanameritenow.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+author: Icantthinkofanameritenow
+
+delete-after: True
+
+changes: 
+- rscdel: Changelings and vampires are no longer able to absorb genomes or suck blood from monkeymen, respectively. Sorry guys, but you gotta work for your powers now.
```
Comment (9600bauds) [inline html/changelogs/icantthinkofanameritenow.yml]: Put this whole line between quotes

**Comment by CptWad** [comment]
there better be a ling overhaul in the future
otherwise :+1: :100:

**Comment by icantthinkofanameritenow** [comment]
Changelog is added, merge when?

**Comment by ihadtoregisterforthis** [comment]
Add "Closes #7637" somewhere.

**Comment by icantthinkofanameritenow** [comment]
Done. Now is it ready?

**Comment by ComicIronic** [comment]
Don't ask for merge, nobody appreciates it and the collabs don't forget that PRs exist so you're not helping yourself.

**Comment by 9600bauds** [comment]
> the collabs don't forget that PRs exist

Funi joek

**Comment by 9600bauds** [comment]
Well this idea has been around for ages and I've never heard a single word against it, plus multiple people guarantee this works, so I see no point to have this wait any longer

**Comment by ihadtoregisterforthis** [comment]
Was coding it yourself so hard?

**Comment by 9600bauds** [comment]
Parkinson's Law of Triviality

**Comment by Probe1** [comment]
@icantthinkofanameritenow 

> Was coding it yourself so hard?

You missed a check to see if a mob had a mind previously.  Now players can go catatonic to spite antags.

**Comment by PJB3005** [comment]
@Probe1 nope not how that works, a mob cannot lose its mind (except for shit like mindswap)

**Comment by Probe1** [comment]
@PJB3005 I hope you're right.

**Comment by Imagebard** [comment]
>Now players can go catatonic to spite antags.

I SPECIFICALLY asked the coders if they were sure this wouldn't happen a few posts ago.

**Comment by Imagebard** [comment]
In retrospect I should just have asked how mind worked, since its such a old feature.
Point is every coder should know how (!T.mind) works by now.

**Comment by icantthinkofanameritenow** [comment]
@Imagebard
And I specifically tested it to make sure that it couldn't happen, which showed that it couldn't. Probe just misunderstands how the mind function works.

**Comment by Imagebard** [comment]
Yes that's what I'm saying.

**Comment by 9600bauds** [comment]
> Every coder should know how [something specific to SS13 code] works by now.

get this spider off me

**Comment by clusterfack** [comment]
I don't expect any coder to know shit about a system unless they refactored or wrote code using the damn thing themselves.

**Pull request merged**